orderer/common/multichannel/registrar_test.go

```diff
@@ -1543,6 +1571,119 @@ func TestRegistrar_RemoveChannel(t *testing.T) {
+			os.RemoveAll(filepath.Join(tmpdir, "pendingops", "remove"))
+			if bytes.Compare(b.Header.DataHash, genesisBlockAppRaft.Header.DataHash) == 0 {
+				return true, nil
+			}
+			return false, nil
+		}
+
+		err := registrar.RemoveChannel("raft-sys-channel")
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "failed removing ledger for channel channel-im-not-a-member-of")
+
+		// After removing the system channel, it no longer exists in the registrar or the ledger
+		require.Empty(t, registrar.SystemChannelID())
+		require.Nil(t, registrar.SystemChannel())
+		require.Nil(t, registrar.GetChain("raft-sys-channel"))
+		require.NotContains(t, ledgerFactory.ChannelIDs(), "raft-sys-channel")
+
+		// channels the orderer wasn't a member no longer exist
+		require.NotNil(t, registrar.GetChain("channel-im-not-a-member-of"))
+		require.Contains(t, ledgerFactory.ChannelIDs(), "channel-im-not-a-member-of")
+	})
+
+	t.Run("app channel is already being removed", func(t *testing.T) {
+		setup(t)
+		defer cleanup()
+
+		consenter.IsChannelMemberReturns(true, nil)
+		registrar := NewRegistrar(config, ledgerFactory, mockCrypto(), &disabled.Provider{}, cryptoProvider, dialer)
+		registrar.Initialize(mockConsenters)
+
+		require.Nil(t, registrar.GetChain("my-raft-channel"))
+		require.NotContains(t, ledgerFactory.ChannelIDs(), "my-raft-channel")
+		info, err := registrar.JoinChannel("my-raft-channel", genesisBlockAppRaft, true)
+		require.NoError(t, err)
+		require.Equal(t, types.ChannelInfo{Name: "my-raft-channel", URL: "", ClusterRelation: "member", Status: "active", Height: 1}, info)
+		require.NotNil(t, registrar.GetChain("my-raft-channel"))
+		require.Contains(t, ledgerFactory.ChannelIDs(), "my-raft-channel")
+
+		registrar.pendingRemoval["my-raft-channel"] = true
+		err = registrar.RemoveChannel("my-raft-channel")
+		require.Error(t, err)
+		require.Equal(t, types.ErrChannelPendingRemoval, err)
+
+		require.Contains(t, registrar.ChannelList().Channels, types.ChannelInfoShort{Name: "my-raft-channel"})
+	})
+
+	t.Run("remove channel fails", func(t *testing.T) {
+		setup(t)
+		defer cleanup()
+
+		consenter.IsChannelMemberReturns(true, nil)
+		registrar := NewRegistrar(config, ledgerFactory, mockCrypto(), &disabled.Provider{}, cryptoProvider, dialer)
+		registrar.Initialize(mockConsenters)
+
+		require.Nil(t, registrar.GetChain("my-raft-channel"))
+		require.NotContains(t, ledgerFactory.ChannelIDs(), "my-raft-channel")
+		info, err := registrar.JoinChannel("my-raft-channel", genesisBlockAppRaft, true)
+		require.NoError(t, err)
+		require.Equal(t, types.ChannelInfo{Name: "my-raft-channel", URL: "", ClusterRelation: "member", Status: "active", Height: 1}, info)
+		require.NotNil(t, registrar.GetChain("my-raft-channel"))
+		require.Contains(t, ledgerFactory.ChannelIDs(), "my-raft-channel")
+
+		// This will force the failure of the underlying factory.Remove() because the pendingops/remove will not exist
```

Comment:
This looks exactly as if the operation was a success. I understand that the failure is asynch, in another thread, so that the call starts ok. However, if the asynch removal failed, the channel is not "gone" from the ledger... and the .remove pendingops file was not deleted, I assume. So maybe we need to keep a channel that its removal failed in the "pendingRemoval" map, so that an attempt to re-join or re-remove would fail, and that it would show up as pending in the List and ListOne?

Reply:
@tock-ibm I decided to use the bool value to determine a "pending" vs "failed" removal. Thoughts?

Reply:
Opened https://jira.hyperledger.org/browse/FAB-18315 as a follow-up